jsrc/algorithm.hpp

```diff
@@ -1,39 +1,119 @@
-
 #include <algorithm>
+#include <array>
 #include <tuple>
 
 namespace algo {
 
+/**
+ * @brief Variadic version of `std::equal_to`
+ */
+static auto constexpr is_equal_to = [](auto const &first, auto const &...values) -> bool {
+    static_assert(sizeof...(values) != 0);
+    return ((first == values) && ...);
+};
+
+/**
+ * @brief Variadic version of `std::not_equal_to`
+ */
+static auto constexpr is_not_equal_to = [](auto const &first, auto const &...values) -> bool {
+    return !is_equal_to(first, values...);
+};
+
 /**
  * @brief Variadic version of `std::mismatch`
  */
 template <typename P, typename I, typename... Is>
-[[nodiscard]] auto
-zip_find(P pred, I f, I l, Is... fs) {
+[[nodiscard]] static auto constexpr zip_find(P const &pred, I f, I l, Is... fs) {
+    static_assert(sizeof...(Is) != 0);
     while (f != l) {
         if (pred(*f, *fs...)) break;
         ++f, ((void)++fs, ...);
     }
     return std::tuple{f, fs...};
 }
 
+/**
+ * @brief Variadic version of `std::mismatch`, for ranges.
+ * @note This can be the same name as zip_find if we use concepts.
+ */
+template <typename P, typename R, typename... Rs>
+[[nodiscard]] static auto constexpr zip_find_r(P const &pred, R const &r, Rs const &...rs) {
+    return zip_find(pred, std::cbegin(r), std::cend(r), std::cbegin(rs)...);
+}
+
 /**
  * @brief Predicate version of `algo::zip_found`
  */
 template <typename P, typename I, typename... Is>
-[[nodiscard]] auto
-zip_found(P pred, I f, I l, Is... fs) {
+[[nodiscard]] static auto constexpr zip_found(P const &pred, I f, I l, Is... fs) -> bool {
     auto const t = zip_find(pred, f, l, fs...);
     return std::get<0>(t) != l;
 }
 
+/**
+ * @brief Variadic version of `algo::zip_found`, for ranges.
+ * @note This can be the same name as zip_found if we use concepts.
+ */
+template <typename P, typename R, typename... Rs>
+[[nodiscard]] static auto constexpr zip_found_r(P const &pred, R const &r, Rs const &...rs) -> bool {
+    return zip_found(pred, std::cbegin(r), std::cend(r), std::cbegin(rs)...);
+}
+
 /**
  * @brief Predicate version of `std::mismatch`
+ * @note std::mismatch isn't constexpr till cpp20
  */
 template <typename I, typename I2>
-[[nodiscard]] auto
-is_mismatched(I f, I l, I2 f2) {
+[[nodiscard]] static auto
+is_mismatched(I f, I l, I2 f2) -> bool {
     return std::mismatch(f, l, f2).first != l;
 }
 
+/**
+ * @brief Variadic, Predicate version of `std::mismatch`
+ */
+template <typename I, typename... Is>
+[[nodiscard]] static auto constexpr is_mismatched_v(I f, I l, Is... fs) -> bool {
```

Comment:
Don't think we need `_v`, function overloading is fine.
```suggestion
[[nodiscard]] static auto constexpr is_mismatched(I f, I l, Is... fs) -> bool {
```

Reply:
I just wasn't sure if I should replace the existing one.